scipy/spatial/transform/_rotation.pyx

```diff
@@ -115,10 +115,72 @@ cdef inline void _quat_canonical(double[:, :] q) noexcept:
     for ind in range(n):
         _quat_canonical_single(q[ind])
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline void _get_angles(
+    double[:] angles, bint extrinsic, bint symmetric, bint sign, 
+    double lamb, double a, double b, double c, double d):
+    
+    # intrinsic/extrinsic conversion helpers
+    cdef int angle_first, angle_third
+    if extrinsic:
+        angle_first = 0
+        angle_third = 2
+    else:
+        angle_first = 2
+        angle_third = 0
+
+    cdef double half_sum, half_diff
+    cdef int case
+
+    # Step 2
+    # Compute second angle...
+    angles[1] = 2 * atan2(hypot(c, d), hypot(a, b))
+
+    # ... and check if equal to is 0 or pi, causing a singularity
+    if abs(angles[1]) <= 1e-7:
+        case = 1
+    elif abs(angles[1] - <double>pi) <= 1e-7:
+        case = 2
+    else:
+        case = 0 # normal case
+
+    # Step 3
+    # compute first and third angles, according to case
+    half_sum = atan2(b, a)
+    half_diff = atan2(d, c)
+    
+    if case == 0:  # no singularities
+        angles[angle_first] = half_sum - half_diff
+        angles[angle_third] = half_sum + half_diff
+    
+    else:  # any degenerate case
+        angles[2] = 0
+        if case == 1:
+            angles[0] = 2 * half_sum
+        else:
+            angles[0] = 2 * half_diff * (-1 if extrinsic else 1)
+            
+    # for Tait-Bryan/asymmetric sequences
+    if not symmetric:
+        angles[angle_third] *= sign
+        angles[1] -= lamb
+
+    for idx in range(3):
+        if angles[idx] < -pi:
+            angles[idx] += 2 * pi
+        elif angles[idx] > pi:
+            angles[idx] -= 2 * pi
+
+    if case != 0:
```

Comment:
I thought that maybe this warning should be raised from the calling function and only once (like if at least one rotation had this property, then the warning is raised after all were processed). 

Basically to reduce overhead of using Python API from here and remove rather useless duplicated warnings.

But as this case is special and rare, it's up to you if you want to change it like this.

Reply:
I think this was already the case though with the `as_euler` function, so maybe this should be changed in a different PR, what do you think?

Moreover, doesn't `warn` already has protections in order to avoid multiple repeated warnings?

Reply:
That's fair.

And I think you are right about suppression of repeated warnings by default. So I guess it's fine to leave it as it is.